snapcraft/plugins/v2/colcon.py

```diff
@@ -0,0 +1,233 @@
+                    "uniqueItems": True,
+                    "items": {"type": "string"},
+                    "default": [],
+                },
+            },
+        }
+
+    def get_build_snaps(self) -> Set[str]:
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        return {
+            "python3-colcon-common-extensions",
+            "python3-rosdep",
+            "python3-wstool",
+            "python3-rosinstall",
+        }
+
+    def get_build_environment(self) -> Dict[str, str]:
+        return {
+            "AMENT_PYTHON_EXECUTABLE": "/usr/bin/python3",
+            "COLCON_PYTHON_EXECUTABLE": "/usr/bin/python3",
+            "ROS_PYTHON_VERSION": "3",
+        }
+
+    def _get_colcon_build_command(self) -> str:
+        cmd: List[str] = [
+            "colcon",
+            "build",
+            "--merge-install",
+            "--install-base",
+            "$SNAPCRAFT_PART_INSTALL",
+        ]
+
+        if self.options.colcon_packages_ignore:
+            cmd.extend(["--packages-ignore", *self.options.colcon_packages_ignore])
+
+        if self.options.colcon_packages:
+            cmd.extend(["--packages-select", *self.options.colcon_packages])
+
+        if self.options.colcon_ament_cmake_args:
+            cmd.extend(["--ament-cmake-args", *self.options.colcon_ament_cmake_args])
+
+        if self.options.colcon_catkin_cmake_args:
+            cmd.extend(["--catkin-cmake-args", *self.options.colcon_catkin_cmake_args])
+
+        # Specify the number of workers
+        cmd.extend(["--parallel-workers", "${SNAPCRAFT_PARALLEL_BUILD_COUNT}"])
+
+        return " ".join(cmd)
+
+    def _get_stage_runtime_dependencies_command(self):
+        env = dict(
+            (key, os.environ[key])
+            for key in ["SNAP", "SNAP_ARCH", "SNAP_NAME", "SNAP_VERSION"]
+            if key in os.environ
+        )
+        env["LC_ALL"] = "C.UTF-8"
+        env["LANG"] = "C.UTF-8"
+
+        return " ".join(
+            _get_python_command(
+                env, [os.path.abspath(__file__), "stage-runtime-dependencies"]
+            )
+        )
+
+    def get_build_commands(self) -> List[str]:
+        return [
+            ". /opt/ros/$ROS_DISTRO/setup.sh",
+            "if [ ! -f /etc/ros/rosdep/sources.list.d/20-default.list ]; then sudo rosdep init; fi",
+            "rosdep update --include-eol-distros --rosdistro $ROS_DISTRO",
+            "rosdep install --from-paths . --default-yes --ignore-packages-from-source",
+            self._get_colcon_build_command(),
+            self._get_stage_runtime_dependencies_command(),
+        ]
+
+
+@click.group()
+def plugin_cli():
+    pass
+
+
+@plugin_cli.command()
+@click.option("--part-install", envvar="SNAPCRAFT_PART_INSTALL", required=True)
+@click.option("--ros-distro", envvar="ROS_DISTRO", required=True)
+def stage_runtime_dependencies(part_install: str, ros_distro: str):
+    # XXX: Do we have to stage python packages, or are they all in the repo?
```

Comment:
Not sure what you're asking here, is this a question that I can help with?

Reply:
I was wondering if all of the ROS packages that are fetched via rosdep are (also) shipped as debs, and if pip support was really required..?

Reply:
Sadly no, some deps are ONLY pip, see [LP: #1683036](https://bugs.launchpad.net/snapcraft/+bug/1683036).

Reply:
Roger that.

Reply:
Note that I'm not opposed to that being a follow-up, personally. This is a great step without it.

Reply:
Agreed to revisit in future.